src/akushon/action/process/interpolator.cpp

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2021 Ichiro ITS
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include <string>
+#include <vector>
+
+#include "akushon/action/process/interpolator.hpp"
+
+#include "akushon/action/model/action.hpp"
+#include "akushon/action/process/joint_process.hpp"
+#include "tachimawari/joint/model/joint.hpp"
+
+namespace akushon
+{
+
+Interpolator::Interpolator(const std::vector<Action> & actions, const Pose & initial_pose)
+: actions(actions), joint_processes({}), current_pose_index(0),
+  pause_time(0), init_pause(false), start_stop_time(0), init_state(true),
+  current_action_index(0)
+{
+  for (const auto & joint : initial_pose.get_joints()) {
+    joint_processes.insert(
+      {joint.get_id(),
+        JointProcess(joint.get_id(), joint.get_position())});
+  }
+
+  if (actions.size() != 0) {
+    state = START_DELAY;
+  } else {
+    state = END;
+  }
+}
+
+void Interpolator::process(int time)
+{
+  switch (state) {
+    case START_DELAY:
+      {
+        if (init_state) {
+          init_state = false;
+          start_stop_time = time;
+        }
+
+        if ((time - start_stop_time) > (actions[current_action_index].get_start_delay() * 1000)) {
+          change_state(PLAYING);
+        }
+
+        break;
+      }
+
+    case PLAYING:
+      {
+        if (check_for_next()) {
+          if (init_pause) {
+            init_pause = false;
+            pause_time = time;
+          }
+
+          if (current_pose_index == actions[current_action_index].get_pose_count()) {
+            change_state(STOP_DELAY);
+            init_pause = true;
+          } else if ((time - pause_time) >  // NOLINT
+            (actions[current_action_index].get_pose(current_pose_index).get_pause() * 1000))
```

Comment:
can we save `actions[current_action_index]` to shorten lines? (and prevent linter error)
```c++
const auto & current_action = actions[current_action_index];
if (current_pose_index == current_action .get_pose_count()) {
  // ...
} else if ((time - pause_time) >  (current_action .get_pose(current_pose_index).get_pause() * 1000)) {
  // ...
}
```
or save it even more
```c++
const auto & current_action = actions[current_action_index];
if (current_pose_index == current_action .get_pose_count()) {
  // ...
} else {
  const auto & current_pose = current_action .get_pose(current_pose_index)
  if ((time - pause_time) >  (current_pose.get_pause() * 1000)) {
    // ...
  }
}
```